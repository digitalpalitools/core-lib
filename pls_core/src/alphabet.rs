use core::str::Chars;
use num_enum::{IntoPrimitive, TryFromPrimitive};
use std::{convert::TryFrom, iter::Peekable};

// TODO: Split this into a Pāli core and a Roman specific module.

// Spec: https://docs.google.com/document/d/1KF6NLFiiVH9oVz_NcU5mjHcMcIAZECgNifM8mX25MCo/edit#heading=h.2hvqs8bpra4
#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, IntoPrimitive, TryFromPrimitive)]
#[repr(usize)]
pub enum PaliAlphabet {
    A,
    Aa,
    I,
    Ii,
    U,
    Uu,
    E,
    O, // vowels - 0-7
    K,
    Kh,
    G,
    Gh,
    QuoteN, // guttural - 8-12
    C,
    Ch,
    J,
    Jh,
    TildeN, // palatal - 13-17
    DotT,
    DotTH,
    DotD,
    DotDH,
    DotN, // retroflex cerebral - 18-22
    T,
    Th,
    D,
    Dh,
    N, // dental - 23-27
    P,
    Ph,
    B,
    Bh,
    M, // labial - 28-32
    Y,
    R,
    L,
    V,
    S,
    H,
    DotL, // semi-vowel - 33-39
    DotM, // nigahita - 40-40
}

pub const PALI_ALPHABET_ROMAN: &[&str] = &[
    "a", "ā", "i", "ī", "u", "ū", "e", "o", // vowels - 0-7
    "k", "kh", "g", "gh", "ṅ", // guttural - 8-12
    "c", "ch", "j", "jh", "ñ", // palatal - 13-17
    "ṭ", "ṭh", "ḍ", "ḍh", "ṇ", // retroflex cerebral - 18-22
    "t", "th", "d", "dh", "n", // dental - 23-27
    "p", "ph", "b", "bh", "m", // labial - 28-32
    "y", "r", "l", "v", "s", "h", "ḷ", // semi-vowel - 33-39
    "ṃ", // nigahita - 40-40
];

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Character {
    Pali(PaliAlphabet),
    Other(char),
}

pub struct CharacterTokenizer<'a> {
    source: Peekable<Chars<'a>>,
}

pub fn char_compare(c1: Character, c2: Character) -> isize {
    match (c1, c2) {
        (Character::Other(c1), Character::Other(c2)) => (c1 as isize - c2 as isize).signum(),
        (Character::Pali(c1), Character::Pali(c2)) => (c1 as isize - c2 as isize).signum(),
        (Character::Other(_c1), Character::Pali(_c2)) => 1,
        (Character::Pali(_c1), Character::Other(_c2)) => -1,
    }
}

pub fn string_compare(str1: &str, str2: &str) -> isize {
    let chars1 = CharacterTokenizer::new(str1.chars());
    let chars2 = CharacterTokenizer::new(str2.chars());

    let cmp = chars1
        .zip(chars2)
        .map(|(c1, c2)| char_compare(c1, c2))
        .find(|&sn| sn != 0);

    match cmp {
        Some(cmp) => cmp,
        None => {
            // TODO: This is a temp hack. Tokenize just once.
            let str1len = string_length(str1);
            let str2len = string_length(str2);
            if str1len != str2len {
                (isize::try_from(str1len).unwrap() - isize::try_from(str2len).unwrap()).signum()
            } else {
                0isize
            }
        }
    }
}

pub fn string_length(str1: &str) -> usize {
    let chars1 = CharacterTokenizer::new(str1.chars());

    chars1.count()
}

impl<'a> CharacterTokenizer<'a> {
    pub fn new(source: Chars<'a>) -> CharacterTokenizer<'a> {
        CharacterTokenizer {
            source: source.peekable(),
        }
    }
}

fn parse_multichar_letter(
    chars: &mut Peekable<Chars<'_>>,
    a1: PaliAlphabet,
    a2: PaliAlphabet,
) -> Character {
    match chars.peek() {
        Some('h') => {
            chars.next();
            Character::Pali(a2)
        }
        _ => Character::Pali(a1),
    }
}

fn parse_singlechar_letter(a: PaliAlphabet) -> Character {
    Character::Pali(a)
}

impl<'a> Iterator for CharacterTokenizer<'a> {
    type Item = Character;

    fn next(&mut self) -> Option<Character> {
        match self.source.next() {
            Some('a') => Some(parse_singlechar_letter(PaliAlphabet::A)),
            Some('ā') => Some(parse_singlechar_letter(PaliAlphabet::AA)),
            Some('i') => Some(parse_singlechar_letter(PaliAlphabet::I)),
            Some('ī') => Some(parse_singlechar_letter(PaliAlphabet::II)),
            Some('u') => Some(parse_singlechar_letter(PaliAlphabet::U)),
            Some('ū') => Some(parse_singlechar_letter(PaliAlphabet::UU)),
            Some('e') => Some(parse_singlechar_letter(PaliAlphabet::E)),
            Some('o') => Some(parse_singlechar_letter(PaliAlphabet::O)),
            Some('k') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::K,
                PaliAlphabet::KH,
            )),
            Some('g') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::G,
                PaliAlphabet::GH,
            )),
            Some('ṅ') => Some(parse_singlechar_letter(PaliAlphabet::QuoteN)),
            Some('c') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::C,
                PaliAlphabet::CH,
            )),
            Some('j') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::J,
                PaliAlphabet::JH,
            )),
            Some('ñ') => Some(parse_singlechar_letter(PaliAlphabet::TildeN)),
            Some('ṭ') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::DotT,
                PaliAlphabet::DotTH,
            )),
            Some('ḍ') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::DotD,
                PaliAlphabet::DotDH,
            )),
            Some('ṇ') => Some(parse_singlechar_letter(PaliAlphabet::DotN)),
            Some('t') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::T,
                PaliAlphabet::TH,
            )),
            Some('d') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::D,
                PaliAlphabet::DH,
            )),
            Some('n') => Some(parse_singlechar_letter(PaliAlphabet::N)),
            Some('p') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::P,
                PaliAlphabet::PH,
            )),
            Some('b') => Some(parse_multichar_letter(
                &mut self.source,
                PaliAlphabet::B,
                PaliAlphabet::BH,
            )),
            Some('m') => Some(parse_singlechar_letter(PaliAlphabet::M)),
            Some('y') => Some(parse_singlechar_letter(PaliAlphabet::Y)),
            Some('r') => Some(parse_singlechar_letter(PaliAlphabet::R)),
            Some('l') => Some(parse_singlechar_letter(PaliAlphabet::L)),
            Some('v') => Some(parse_singlechar_letter(PaliAlphabet::V)),
            Some('s') => Some(parse_singlechar_letter(PaliAlphabet::S)),
            Some('h') => Some(parse_singlechar_letter(PaliAlphabet::H)),
            Some('ḷ') => Some(parse_singlechar_letter(PaliAlphabet::DotL)),
            Some('ṃ') => Some(parse_singlechar_letter(PaliAlphabet::DotM)),
            Some(c) => Some(Character::Other(c)),
            None => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;
    use std::convert::TryFrom;
    use test_case::test_case;

    #[test]
    fn test_pali_alphabet_length() {
        assert_eq!(PALI_ALPHABET_ROMAN.len(), 41);

        let i: usize = PaliAlphabet::DotM.into();
        assert_eq!(i, 40);
    }

    const PALI_ALPHABET_ROMAN_COMPOUND_LETTERS_INDICES: &[usize] =
        &[8, 10, 13, 15, 18, 20, 23, 25, 28, 30];

    const PALI_ALPHABET_ROMAN_COMPOUNDING_LETTER_INDEX: usize = 38;

    fn is_compound_letter_roman(index: usize) -> bool {
        None != PALI_ALPHABET_ROMAN_COMPOUND_LETTERS_INDICES
            .iter()
            .find(|&&e| e == index)
    }

    fn fixup_compound_letters(indices: &[usize]) -> Vec<usize> {
        let ret = indices
            .iter()
            .fold((None, Vec::new()), |(ep, mut acc), &e| {
                match ep {
                    Some(ep) => {
                        if e == PALI_ALPHABET_ROMAN_COMPOUNDING_LETTER_INDEX
                            && is_compound_letter_roman(ep)
                        {
                            let i_prev = acc.len() - 1;
                            acc[i_prev] += 1;
                        } else {
                            acc.push(e);
                        }
                    }
                    None => acc.push(e),
                }
                (Some(e), acc)
            });

        ret.1
    }

    #[test]
    fn parse_with_non_pali_character() {
        let str = "xā1b";
        let tokenizer = CharacterTokenizer::new(str.chars());

        let chars: Vec<_> = tokenizer.collect();

        assert_eq!(
            chars,
            vec![
                Character::Other('x'),
                Character::Pali(PaliAlphabet::AA),
                Character::Other('1'),
                Character::Pali(PaliAlphabet::B)
            ]
        )
    }

    #[test]
    fn fixup_compound_letters_with_no_compound_letters() {
        let indices: Vec<usize> = vec![0, 1, 2];
        let fixedup_indices = fixup_compound_letters(&indices);

        assert_eq!(fixedup_indices, indices)
    }

    #[test_case("c", "cc"   => -1)]
    #[test_case("c", "b"    => -1)]
    #[test_case("c", "c"    => 0)]
    #[test_case("b", "c"    => 1)]
    #[test_case("cc", "c"   => 1)]
    #[test_case("ac", "ab"  => -1)]
    #[test_case("ac", "ac"  => 0)]
    #[test_case("ab", "ac"  => 1)]
    #[test_case("a", "x"    => -1)]
    #[test_case("x", "a"    => 1)]
    #[test_case("x", "z"    => -1)]
    #[test_case("x", "x"    => 0)]
    #[test_case("z", "x"    => 1)]
    #[test_case("xabc", "aabc"  => 1)]
    #[test_case("aabc", "xabc"  => -1)]
    #[test_case("xabc", "yabc"  => 1)]
    #[test_case("xabc", "xabc"  => 0)]
    #[test_case("yabc", "xabc"  => -1)]
    #[test_case("i", "ā"    => 1; "random letters 1")]
    #[test_case("cc", "b"   => -1; "longer of lesser sort order 1")]
    fn string_compare_tests(str1: &str, str2: &str) -> isize {
        string_compare(str1, str2)
    }

    #[test_case("buddho" => 5usize; "simple word 1")]
    #[test_case("bhagavā" => 6usize; "simple word 2")]
    #[test_case("aāiīuūeokkhgghṅcchjjhñṭṭhḍḍhṇtthddhnpphbbhmyrlvshḷṃ" => 41; "all characters")]
    fn string_length_tests(str1: &str) -> usize {
        string_length(str1)
    }

    proptest! {
        #[test]
        fn string_compare_all(i1 in 0usize..PALI_ALPHABET_ROMAN.len(), i2 in 0usize..PALI_ALPHABET_ROMAN.len()) {
            let cmp_str = string_compare(PALI_ALPHABET_ROMAN[i1], PALI_ALPHABET_ROMAN[i2]);

            let pali_char1 = PaliAlphabet::try_from(i1).expect("Catastrophic unhandlable error.");
            let pali_char2 = PaliAlphabet::try_from(i2).expect("Catastrophic unhandlable error.");
            let cmp_char = (pali_char1 as isize - pali_char2 as isize).signum();

            assert_eq!(cmp_char, cmp_str);
        }

        #[test]
        fn fixup_compound_letters_with_compound_letters(index in 0usize..PALI_ALPHABET_ROMAN_COMPOUND_LETTERS_INDICES.len()) {
            let indices: Vec<usize> = vec![0, PALI_ALPHABET_ROMAN_COMPOUND_LETTERS_INDICES[index], 38, 38, 2, 38];
            let fixed_indices = fixup_compound_letters(&indices);

            let new_indices = vec![0, PALI_ALPHABET_ROMAN_COMPOUND_LETTERS_INDICES[index] + 1, 38, 2, 38];

            assert_eq!(new_indices, fixed_indices)
        }

        #[test]
        fn round_trip_pali_to_roman(index in 0usize..PALI_ALPHABET_ROMAN.len()) {
            let pali_char = PaliAlphabet::try_from(index).expect("Catastrophic unhandlable error.");
            let i: usize = pali_char.into();
            let str = PALI_ALPHABET_ROMAN[i];

            let tokenizer = CharacterTokenizer::new(str.chars());
            let new_pali_char = tokenizer.map(|c| match c { Character::Pali(c) => c, _ => panic!("") }).next().expect("Catastrophic unhandlable error.");

            assert_eq!(new_pali_char, pali_char);
        }

        #[test]
        fn round_trip_parsing_for_long_strings(indices in prop::collection::vec(0usize..PALI_ALPHABET_ROMAN.len(), 0..100)) {
            let indices = fixup_compound_letters(&indices);

            let pali_string = indices
                .iter()
                .map(|&i| PALI_ALPHABET_ROMAN[i] )
                .fold(String::new(), |acc, e| acc + e);

            let tokenizer = CharacterTokenizer::new(pali_string.chars());
            let new_indices: Vec<usize> = tokenizer.map(|c| match c { Character::Pali(c) => c.into(), _ => panic!("") }).collect();

            assert_eq!(new_indices, indices);
        }
    }
}
