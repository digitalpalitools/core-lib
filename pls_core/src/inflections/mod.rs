mod generators;

use crate::alphabet::string_compare;
use regex::{Error, Regex};
use serde::Serialize;
use std::collections::HashMap;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![("output", include_str!("templates/output.html"))])
            .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
    static ref INDECLINABLE_CRACKER: Result<Regex, Error> = Regex::new(r" \d+$");
}

#[derive(Debug)]
pub enum InflectionClass {
    Indeclinable,
    Conjugation,
    Declension,
    DeclensionPron1st,
    DeclensionPron2nd,
    DeclensionPronDual,
}

pub struct Pali1Metadata {
    pub pali1: String,
    pub stem: String,
    pub pattern: String,
    pub pos: String,
    pub meaning: String,
    pub inflection_class: InflectionClass,
    pub like: String,
}

pub struct SqlQuery {
    exec_sql_query: fn(&str) -> Result<String, String>,
}

impl SqlQuery {
    fn new(exec_sql_query: fn(&str) -> Result<String, String>) -> SqlQuery {
        SqlQuery { exec_sql_query }
    }

    fn exec(&self, query: &str) -> Result<Vec<Vec<Vec<String>>>, String> {
        let result_str = (self.exec_sql_query)(&query)?;
        let result: Vec<Vec<Vec<String>>> =
            serde_json::from_str(&result_str).map_err(|e| e.to_string())?;
        Ok(result)
    }
}

pub fn generate_inflection_table(
    pali1: &str,
    host_url: &str,
    host_version: &str,
    transliterate: fn(&str) -> Result<String, String>,
    exec_sql_query: fn(&str) -> Result<String, String>,
    locale: &str,
) -> Result<String, String> {
    let q = SqlQuery::new(exec_sql_query);
    let pm = get_pali1_metadata(pali1, transliterate, &q)?;
    let body = generators::create_html_body(&pm, transliterate, &q, locale)?;

    generate_output(&pm, pali1, host_url, host_version, &body, transliterate)
}

fn inflection_class_from_str(ic: &str) -> InflectionClass {
    match ic {
        "verb" => InflectionClass::Conjugation,
        "pron1st" => InflectionClass::DeclensionPron1st,
        "pron2nd" => InflectionClass::DeclensionPron2nd,
        "prondual" => InflectionClass::DeclensionPronDual,
        _ => InflectionClass::Declension,
    }
}

fn get_stem_for_indeclinable(pali1: &str) -> Result<String, String> {
    let regex = INDECLINABLE_CRACKER.as_ref().map_err(|e| e.to_string())?;
    Ok(regex.replace(pali1, "").to_string())
}

fn get_pali1_metadata(
    pali1: &str,
    transliterate: fn(&str) -> Result<String, String>,
    q: &SqlQuery,
) -> Result<Pali1Metadata, String> {
    let sql = format!(
        r#"select stem, pattern, pos, definition from '_stems' where pƒÅli1 = "{}""#,
        pali1,
    );
    let results = q.exec(&sql)?;
    let stem = &results[0][0][0];
    let pattern = &results[0][0][1];
    let mut pm = Pali1Metadata {
        pali1: pali1.to_string(),
        stem: if !stem.eq("*") {
            stem.to_owned()
        } else {
            "".to_string()
        },
        pattern: pattern.to_owned(),
        pos: results[0][0][2].to_owned(),
        meaning: results[0][0][3].to_owned(),
        inflection_class: InflectionClass::Declension,
        like: "".to_string(),
    };

    if !pattern.trim().is_empty() {
        let sql = format!(
            r#"select inflection_class, like from '_index' where name = "{}""#,
            pattern
        );
        let results = q.exec(&sql)?;
        let inflection_class = &results[0][0][0];
        let like = &results[0][0][1];

        pm.inflection_class = inflection_class_from_str(inflection_class);
        pm.like = if !like.is_empty() {
            format!("like {}", transliterate(like)?)
        } else {
            "irregular".to_string()
        };
    } else if stem.eq("-") {
        pm.inflection_class = InflectionClass::Indeclinable;
        pm.pali1 = get_stem_for_indeclinable(&pm.pali1)?;
        pm.like = "indeclinable".to_string();
    }

    Ok(pm)
}

#[derive(Serialize)]
struct ViewModel<'a> {
    pub pali1: &'a str,
    pub pattern: &'a str,
    pub like: &'a str,
    pub pos: &'a str,
    pub meaning: &'a str,
    pub body: &'a str,
    pub feedback_form_url: &'a str,
    pub host_url: &'a str,
    pub host_version: &'a str,
}

fn generate_output(
    pm: &Pali1Metadata,
    pali1: &str,
    host_url: &str,
    host_version: &str,
    body: &str,
    transliterate: fn(&str) -> Result<String, String>,
) -> Result<String, String> {
    let feedback_form_url = match pm.inflection_class {
        InflectionClass::Conjugation => {
            "https://docs.google.com/forms/d/e/1FAIpQLSeJpx7TsISkYEXzxvbBtOH25T-ZO1Z5NFdujO5SD9qcAH_i1A/viewform"
        }
        _ => { // All declensions.
            "https://docs.google.com/forms/d/e/1FAIpQLSeoxZiqvIWadaLeuXF4f44NCqEn49-B8KNbSvNer5jxgRYdtQ/viewform"
        }
    };

    let vm = ViewModel {
        pali1: &transliterate(pali1)?,
        pattern: &pm.pattern,
        like: &pm.like,
        pos: &pm.pos,
        meaning: &pm.meaning,
        body: &body,
        feedback_form_url: &feedback_form_url,
        host_url: &host_url,
        host_version: &host_version,
    };

    let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
    TEMPLATES
        .render("output", &context)
        .map_err(|e| e.to_string())
}

#[derive(Debug, Clone)]
pub struct InflectedWordMetadata {
    pub inflected_word: String,
    pub stem_word: String,
    pub grammar: String,
    pub comment: String,
}

impl InflectedWordMetadata {
    pub fn simple_representation(self) -> (String, String, String, String) {
        (
            self.inflected_word,
            self.stem_word,
            self.grammar,
            self.comment,
        )
    }
}

fn get_inflections_for_pattern(
    pattern: &str,
    q: &SqlQuery,
) -> Result<Vec<Vec<Vec<String>>>, String> {
    q.exec(&format!("Select * from {}", pattern))
}

fn get_words_for_indeclinable_stem(pali1: &str) -> Result<Vec<InflectedWordMetadata>, String> {
    Ok(vec![InflectedWordMetadata {
        inflected_word: get_stem_for_indeclinable(pali1)?,
        stem_word: pali1.to_string(),
        grammar: " ".to_string(),
        comment: "ind".to_string(),
    }])
}

fn get_words_for_irregular_stem(
    pali1: &str,
    pattern: &str,
    q: &SqlQuery,
) -> Result<Vec<InflectedWordMetadata>, String> {
    let inflections: Vec<Vec<String>> = get_inflections_for_pattern(pattern, &q)?
        .pop()
        .ok_or_else(|| format!("No pattern found for {}", pattern))?;
    let mut inflected_words_irregular_stem: Vec<InflectedWordMetadata> = Vec::new();
    for mut inflection_row in inflections {
        for inflection in inflection_row
            .pop()
            .ok_or_else(|| format!("No pattern found for {}", pattern))?
            .split(',')
        {
            inflected_words_irregular_stem.push(InflectedWordMetadata {
                inflected_word: inflection.to_string(),
                stem_word: pali1.to_string(),
                grammar: inflection_row.join(" ").to_string(),
                comment: "*".to_string(),
            })
        }
    }
    Ok(inflected_words_irregular_stem)
}

fn get_words_for_regular_stem(
    pali1: &str,
    stem: &str,
    pattern: &str,
    q: &SqlQuery,
) -> Result<Vec<InflectedWordMetadata>, String> {
    let mut inflected_words_regular_stem: Vec<InflectedWordMetadata> = Vec::new();
    let inflections: Vec<Vec<String>> = get_inflections_for_pattern(pattern, &q)?
        .pop()
        .ok_or_else(|| format!("No pattern found for {}", pattern))?;
    for mut inflection_row in inflections {
        for inflection in inflection_row
            .pop()
            .ok_or_else(|| format!("No pattern found for {}", pattern))?
            .split(',')
        {
            inflected_words_regular_stem.push(InflectedWordMetadata {
                inflected_word: [stem, inflection].join("").to_string(),
                stem_word: pali1.to_string(),
                grammar: inflection_row.join(" ").to_string(),
                comment: " ".to_string(),
            })
        }
    }
    Ok(inflected_words_regular_stem)
}

pub fn generate_all_inflected_words(
    pali1: &str,
    stem: &str,
    pattern: &str,
    exec_sql_query: fn(&str) -> Result<String, String>,
) -> Result<Vec<InflectedWordMetadata>, String> {
    let q = SqlQuery::new(exec_sql_query);
    let inflected_words: Vec<InflectedWordMetadata> = match stem {
        "-" => get_words_for_indeclinable_stem(pali1)?,
        "*" => get_words_for_irregular_stem(pali1, pattern, &q)?,
        _ => get_words_for_regular_stem(pali1, stem, pattern, &q)?,
    };
    Ok(inflected_words)
}

fn join_and_transliterate_if_not_empty(
    stem: &str,
    suffix: &str,
    transliterate: fn(&str) -> Result<String, String>,
) -> String {
    if suffix.is_empty() {
        "".to_string()
    } else {
        transliterate(&format!("{}{}", stem, suffix)).unwrap_or_else(|e| e)
    }
}

fn get_inflections(
    stem: &str,
    sql: &str,
    transliterate: fn(&str) -> Result<String, String>,
    q: &SqlQuery,
) -> Vec<String> {
    let res = match q.exec(&sql) {
        Ok(x) => {
            if x.len() == 1 && x[0].len() == 1 && x[0][0].len() == 1 {
                x[0][0][0].to_string()
            } else {
                "".to_string()
            }
        }
        Err(e) => e,
    };

    let mut inflections: Vec<String> = res
        .split(',')
        .map(|s| join_and_transliterate_if_not_empty(stem, s, transliterate))
        .collect();
    inflections.sort_by(|a, b| Ord::cmp(&string_compare(a, b), &0));
    inflections
}

fn query_has_no_results(query: &str, q: &SqlQuery) -> Result<bool, String> {
    let count = &q.exec(&query)?[0][0][0];
    Ok(count.eq("0"))
}

pub fn get_abbreviations_for_locale(
    locale: &str,
    q: &SqlQuery,
) -> Result<HashMap<String, String>, String> {
    let sql: String;
    if locale == "xx" {
        sql = "select name, description, '^' || name || '$' from _abbreviations".to_string();
    } else if locale == "en" {
        sql = "select name, description, name from _abbreviations".to_string();
    } else {
        sql = format!(
            r#"select name, description, {} from _abbreviations"#,
            locale
        );
    }
    let res = q.exec(&sql)?;
    let mut abbrev_map = HashMap::new();
    for i in res[0].iter() {
        abbrev_map.insert(i[0].clone(), i[2].clone());
        abbrev_map.insert(i[1].clone(), i[2].clone());
    }
    let list_of_required_keys = [
        "nom", "acc", "instr", "dat", "abl", "gen", "loc", "voc", "in comps", "masc", "fem", "nt",
        "x", "sg", "pl", "dual", "act", "reflx", "pr", "fut", "aor", "opt", "imp", "cond",
        "imperf", "perf", "1st", "2nd", "3rd", "irreg", "gram", "ind", "abs", "adj", "adv", "base",
        "card", "case", "comp", "cs", "ger", "idiom", "inf", "like", "ordin", "person", "pp",
        "prefix", "pron", "prp", "ptp", "root", "sandhi", "suffix", "ve",
    ];
    for i in list_of_required_keys.iter() {
        if !abbrev_map.contains_key(&i.to_string()) {
            abbrev_map.insert(i.to_string(), format!("{} ABBREVIATION NOT FOUND", i));
        }
    }
    Ok(abbrev_map)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rusqlite::{Connection, Row, NO_PARAMS};
    use test_case::test_case;

    fn get_row_cells(row: &Row) -> Vec<String> {
        let cells: Vec<String> = row
            .column_names()
            .iter()
            .map(|&cn| {
                let cell: String = match row.get(cn) {
                    Ok(c) => c,
                    Err(e) => e.to_string(),
                };
                cell
            })
            .collect();

        cells
    }

    fn exec_sql_core(sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, rusqlite::Error> {
        let conn = Connection::open("../inflections.db")?;
        let mut result: Vec<Vec<Vec<String>>> = Vec::new();
        for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
            let mut stmt = conn.prepare(&s)?;
            let mut rows = stmt.query(NO_PARAMS)?;

            let mut table: Vec<Vec<String>> = Vec::new();
            while let Some(row) = rows.next()? {
                table.push(get_row_cells(row));
            }
            result.push(table)
        }

        Ok(result)
    }

    fn exec_sql(sql: &str) -> Result<String, String> {
        let table = exec_sql_core(&sql).map_err(|x| x.to_string())?;
        serde_json::to_string(&table).map_err(|x| x.to_string())
    }

    fn psuedo_transliterate(s: &str) -> String {
        format!("^{}$", s)
    }

    #[test_case("ƒÅbƒÅdheti","xx"; "conjugation - 1 - xx")]
    #[test_case("vass≈´panƒÅyikƒÅ","xx"; "declension - 1 - xx ")]
    #[test_case("kamma 1","xx"; "declension - 2 - irreg - xx")]
    #[test_case("kƒÅma·πÉ 3","xx"; "declension - 3 - ind - xx")]
    #[test_case("ubha","xx"; "declension - 4 - pron_dual - xx")]
    #[test_case("ma·πÉ","xx"; "declension - 4 - pron_1st - xx")]
    #[test_case("ta·πÉ 3","xx"; "declension - 4 - pron_2nd - xx")]
    #[test_case("pa√±ca","xx"; "declension - 5 - only x gender - xx")]
    #[test_case("ƒÅbƒÅdheti","en"; "conjugation - 1 - en")]
    fn inflection_tests(pali1: &str, locale: &str) {
        let html = generate_inflection_table(
            pali1,
            "test case",
            "v0.1",
            |s| Ok(psuedo_transliterate(s)),
            exec_sql,
            locale,
        )
        .unwrap_or_else(|e| e);
        insta::assert_snapshot!(html);
    }

    #[test_case("a 1", "-", ""; "indeclinable")]
    #[test_case("ababa 1", "abab", "a_nt"; "regular")]
    #[test_case("ahesu·πÉ", "*", "ahosi_aor"; "irregular")]
    fn inflected_word_tests(pali1: &str, stem: &str, pattern: &str) {
        let output: Vec<(String, String, String, String)> =
            generate_all_inflected_words(pali1, stem, pattern, exec_sql)
                .unwrap_or_else(|_e| Vec::new())
                .iter_mut()
                .map(|x| x.clone().simple_representation())
                .collect();

        insta::assert_yaml_snapshot!(output);
    }
}
