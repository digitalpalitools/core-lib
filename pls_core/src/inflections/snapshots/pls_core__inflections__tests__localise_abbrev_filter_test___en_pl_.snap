---
source: pls_core/src/inflections/mod.rs
expression: html

---
pl

